velox/dwio/dwrf/reader/SelectiveColumnReader.cpp

```diff
@@ -3861,6 +3862,294 @@ void SelectiveStringDictionaryColumnReader::ensureInitialized() {
   initTimeClocks_ = timer.elapsedClocks();
 }
 
+template <typename T, typename TFilter, typename ExtractValues, bool isDense>
+class DirectRleColumnVisitor
+    : public ColumnVisitor<T, TFilter, ExtractValues, isDense> {
+  using super = ColumnVisitor<T, TFilter, ExtractValues, isDense>;
+
+ public:
+  DirectRleColumnVisitor(
+      TFilter& filter,
+      SelectiveColumnReader* reader,
+      RowSet rows,
+      ExtractValues values)
+      : ColumnVisitor<T, TFilter, ExtractValues, isDense>(
+            filter,
+            reader,
+            rows,
+            values) {}
+
+  // Use for replacing all rows with non-null rows for fast path with
+  // processRun and processRle.
+  void setRows(folly::Range<const int32_t*> newRows) {
+    super::rows_ = newRows.data();
+    super::numRows_ = newRows.size();
+  }
+
+  // Processes 'numInput' T's in 'input'. Sets 'values' and
+  // 'numValues'' to the resulting values. 'scatterRows' may be
+  // non-null if there is no filter and the decoded values should be
+  // scattered into values with gaps in between so as to leave gaps
+  // for nulls. If scatterRows is given, the ith value goes to
+  // values[scatterRows[i]], else it goes to 'values[i]'. If
+  // 'hasFilter' is true, the passing values are written to
+  // consecutive places in 'values'.
+  template <bool hasFilter, bool hasHook, bool scatter>
+  void processRun(
+      const T* input,
+      int32_t numInput,
+      const int32_t* scatterRows,
+      int32_t* filterHits,
+      T* values,
+      int32_t& numValues) {
+    DCHECK_EQ(input, values + numValues);
+    constexpr bool filterOnly =
+        std::is_same<typename super::Extract, DropValues>::value;
+
+    processFixedWidthRun<T, filterOnly, scatter, isDense>(
+        folly::Range<const vector_size_t*>(super::rows_, super::numRows_),
+        super::rowIndex_,
+        numInput,
+        scatterRows,
+        values,
+        filterHits,
+        numValues,
+        super::filter_,
+        super::values_.hook());
+
+    super::rowIndex_ += numInput;
+  }
+
+  template <bool hasFilter, bool hasHook, bool scatter>
+  void processRle(
```

Comment:
I'd assume similar thing would have already existed. Maybe it's because RLE is applied to like lengths of string, indices into dictionary and it's the first time we have filter/hook for in timestamp?